onnxruntime/core/providers/cuda/nn/conv.cc

```diff
@@ -34,6 +34,52 @@ REGISTER_KERNEL_TYPED(float)
 REGISTER_KERNEL_TYPED(double)
 REGISTER_KERNEL_TYPED(MLFloat16)
 
+template <typename T>
+const cudnnConvolutionFwdAlgo_t Conv<T>::kAllAlgos[] = {
+    CUDNN_CONVOLUTION_FWD_ALGO_GEMM,
+    CUDNN_CONVOLUTION_FWD_ALGO_FFT,
+    CUDNN_CONVOLUTION_FWD_ALGO_FFT_TILING,
+    CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_GEMM,
+    CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM,
+    CUDNN_CONVOLUTION_FWD_ALGO_DIRECT,
+    CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD,
+    CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD_NONFUSED,
+};
+
+cudnnStatus_t getWorkspaceSize(const CudnnConvState<cudnnConvolutionFwdAlgoPerf_t>& s,
+                               cudnnConvolutionFwdAlgo_t algo, size_t* sz) {
+  return cudnnGetConvolutionForwardWorkspaceSize(
+      s.handle,
+      s.x_tensor,
+      s.w_desc,
+      s.conv_desc,
+      s.y_tensor,
+      algo,
+      sz);
+}
+
+template <typename algo_t>
+size_t getMaxWorkspaceSize(const CudnnConvState<cudnnConvolutionFwdAlgoPerf_t>& s,
+                           const algo_t* algo, int n_algo) {
+  size_t max_ws_size = 0;
+
+  // TODO: get maximum available size from memory areana
+
+  size_t free, total;
+  CUDA_CALL_THROW(cudaMemGetInfo(&free, &total));
```

Comment:
calling cudaMemGetInfo is not ideal... but our cuda allocator doesn't have a way to return the current available memory reading...